src/gui/processing/qgsprocessingpointcloudexpressionlineedit.cpp

```diff
@@ -0,0 +1,328 @@
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsprocessingpointcloudexpressionlineedit.h"
+#include "qgsgui.h"
+#include "qgsapplication.h"
+#include "qgsfilterlineedit.h"
+#include "qgspointcloudlayer.h"
+#include "qgspointcloudexpression.h"
+
+#include <QHBoxLayout>
+#include <QToolButton>
+#include <QListView>
+#include <QPushButton>
+#include <QMessageBox>
+
+QgsProcessingPointCloudExpressionLineEdit::QgsProcessingPointCloudExpressionLineEdit( QWidget *parent )
+  : QWidget( parent )
+{
+  mLineEdit = new QgsFilterLineEdit();
+  mLineEdit->setSizePolicy( QSizePolicy::Expanding, QSizePolicy::Minimum );
+
+  mButton = new QToolButton();
+  mButton->setSizePolicy( QSizePolicy::Minimum, QSizePolicy::Minimum );
+  mButton->setIcon( QgsApplication::getThemeIcon( QStringLiteral( "/mIconExpression.svg" ) ) );
+  connect( mButton, &QAbstractButton::clicked, this, &QgsProcessingPointCloudExpressionLineEdit::editExpression );
+
+  QHBoxLayout *layout = new QHBoxLayout();
+  layout->setContentsMargins( 0, 0, 0, 0 );
+  layout->addWidget( mLineEdit );
+  layout->addWidget( mButton );
+  setLayout( layout );
+
+  setSizePolicy( QSizePolicy::Expanding, QSizePolicy::Minimum );
+  setFocusProxy( mLineEdit );
+  connect( mLineEdit, &QLineEdit::textChanged, this, static_cast < void ( QgsProcessingPointCloudExpressionLineEdit::* )( const QString & ) > ( &QgsProcessingPointCloudExpressionLineEdit::expressionEdited ) );
+
+  setExpression( expression() );
+}
+
+QgsProcessingPointCloudExpressionLineEdit::~QgsProcessingPointCloudExpressionLineEdit() = default;
+
+void QgsProcessingPointCloudExpressionLineEdit::setLayer( QgsPointCloudLayer *layer )
+{
+  mLayer = layer;
+}
+
+QgsPointCloudLayer *QgsProcessingPointCloudExpressionLineEdit::layer() const
+{
+  return mLayer;
+}
+
+QString QgsProcessingPointCloudExpressionLineEdit::expression() const
+{
+  if ( mLineEdit )
+    return mLineEdit->text();
+
+  return QString();
+}
+
+void QgsProcessingPointCloudExpressionLineEdit::setExpression( const QString &newExpression )
+{
+  if ( mLineEdit )
+    mLineEdit->setText( newExpression );
+}
+
+void QgsProcessingPointCloudExpressionLineEdit::editExpression()
+{
+  const QString currentExpression = expression();
+  QgsProcessingPointCloudExpressionDialog dlg( mLayer );
```

Comment:
it would be SUPER nice to make this open as a QgsPanelWidget. Then we could avoid an unwanted modal dialog blocking things.

Reply:
Would you mind if I do it in a follow up pull-request?